src/gurt/dyn_hash.c

```diff
@@ -0,0 +1,1269 @@
+		    unsigned int ksize, dh_item_t item, bool exclusive)
+{
+	int		rc = 0;
+	uint64_t	siphash;
+	struct dyn_hash	*htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	siphash = gen_siphash(key, ksize);
+	do {
+		rc = do_insert(htable, key, ksize, &item,
+			       siphash, exclusive ? DH_INSERT_EXCLUSIVE :
+			       DH_INSERT_INCLUSIVE);
+	} while (rc == -DER_AGAIN);
+
+	if (rc == 0) {
+		/* set siphash */
+		htable->ht_ops.hop_siphash_set(item, siphash);
+	}
+	return rc;
+}
+
+bool
+dyn_hash_rec_delete(struct d_hash_table *gtable, const void *key,
+		    unsigned int ksize, uint64_t siphash)
+{
+	uint32_t	index;
+	dh_bucket_t	*bucket;
+	struct dyn_hash *htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	if (siphash == 0) {
+		/* generate siphash */
+		siphash = gen_siphash(key, ksize);
+	}
+	index = (uint32_t)(siphash >> htable->ht_shift);
+	htable->ht_write_lock(htable);
+	bucket = htable->ht_vector.data[index];
+	htable->bucket_lock(bucket);
+	return do_delete(gtable, key, ksize, siphash, bucket, index);
+}
+
+bool
+dyn_hash_rec_delete_at(struct d_hash_table *gtable, dh_item_t item)
+{
+	bool		rc = true;
+	void		*key;
+	unsigned int	ksize;
+	uint64_t	siphash;
+	struct dyn_hash *htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	if (!(htable->ht_ops.hop_key_get(item, &key, &ksize))) {
+		D_ERROR("Get key function failed\n");
+		D_GOTO(out, rc = false);
+	}
+	siphash = gen_siphash(key, ksize);
+	rc = dyn_hash_rec_delete(gtable, (const void *)key, ksize, siphash);
+out:
+	return rc;
+}
+
+bool
+dyn_hash_rec_evict(struct d_hash_table *htable, const void *key,
+		   unsigned int ksize)
+{
+	return false;
+}
+
+bool
+dyn_hash_rec_evict_at(struct d_hash_table *htable, dh_item_t item,
+		      uint64_t siphash)
+{
+	return false;
+}
+
+void
+dyn_hash_rec_addref(struct d_hash_table *gtable, dh_item_t item)
+{
+	struct dyn_hash *htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+	htable->ht_write_lock(htable);
+	htable->ht_ops.hop_rec_addref(gtable, item);
+	htable->ht_rw_unlock(htable);
+}
+
+void
+dyn_hash_rec_decref(struct d_hash_table *gtable, dh_item_t item)
+{
+	struct dyn_hash *htable = gtable->dyn_hash;
+	bool		zombie;
+	bool		ephemeral = (gtable->ht_feats & D_HASH_FT_EPHEMERAL);
+	void		*key;
+	unsigned int	ksize;
+	uint64_t	siphash;
+	uint32_t	index;
+	dh_bucket_t	*bucket;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	htable->ht_write_lock(htable);
+	zombie = htable->ht_ops.hop_rec_decref(gtable, item);
+	if (zombie) {
+		if (!htable->ht_ops.hop_key_get(item, &key, &ksize)) {
+			D_ERROR("Get key function failed\n");
```

Comment:
error should print more information which would be useful in debugging it, such as perhaps gtable and item